stdenv: add applyPatches and support patch directories

tie opened the pull request:
## Description of changes

This PR adds applyPatches function to the stdenv that supports patch directories and refactors the default `patchPhase` to use it. Also documents `applyPatches` trivial builder in Nixpkgs manual.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tie committed:
stdenv: add applyPatches and support patch directories

philiptaron commented:
I'm of two minds about this PR.

On the one hand, I like the way that applyPatches is separated from the patchPhase. I like the way it's upgraded with arrays and all.

On the other hand, as a reviewer, I don't like how it's harder to validate which patches get applied. I _like_ that I can be assured that adding a patch means I get a cue in the `.nix` file.

Adding a patch file and "magically" having it be applied feels not great to me.

Maybe my intuition is off here. I'd love to here from others.

tie commented:
Damn, I really have a bad luck for encountering bugs in every piece of software I touch… I’m adding tests for an existing applyPatches trivial builder (and applyPatches function by extension) and the following test case fails (as expected) because `patch` segfaults (*not* expected but the test succeeds because we expect failure).

<details>

```nix
  patchWithIncorrectFilePath = testers.testBuildFailure (applyPatches {
    src = writeTextDir "file1.txt" "content";
    patches = [
      (writeText "incorrect-path.patch" ''
        diff --git a/file2.txt b/file2.txt
        index e69de29..d20e9cd 100644
        --- a/file2.txt
        +++ b/file2.txt
        @@ -1 +1 @@
        -old content
        +new content
      '')
    ];
  })
```

Standalone reproducer:
```console
$ uname -a
Linux utm 6.6.32 #1-NixOS SMP Sat May 25 14:22:56 UTC 2024 aarch64 GNU/Linux
```
```console
$ uname -a
Linux saitama 6.6.32 #1-NixOS SMP PREEMPT_DYNAMIC Sat May 25 14:22:56 UTC 2024 x86_64 GNU/Linux
```
```console
$ patch --version
GNU patch 2.7.6
(GNU copyright notice omitted)
```
```console
$ patch -p1 <<'EOF'
diff --git a/file2.txt b/file2.txt
index e69de29..d20e9cd 100644
--- a/file2.txt
+++ b/file2.txt
@@ -1 +1 @@
-old content
+new content
EOF
```
Spams
```
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.origSegmentation fault (core dumped)
```

To be fair, the patch file for this test case was generated by LLM, and `patch` doesn’t crash without Git headers.
</details>

tie commented:
>On the other hand, as a reviewer, I don't like how it's harder to validate which patches get applied. I like that I can be assured that adding a patch means I get a cue in the .nix file.

An existing patch file can be updated to add more changes without updating package’s `.nix` file, similar to adding a file in a directory. With patch set directories, Nix files don’t contain a complete list of patch files, but I don’t think it’s really useful without actual patch contents and context, and thus encourages duplication between patch message and comments above patch paths.

Patch directories are useful for managing patches using tools like `git` (`am`, `format-patch`) and [quilt](https://savannah.nongnu.org/projects/quilt). Unless `series` file exists, I think it’s easier to validate which patches are applied: _all_ patches in a directory (with `.patch` extension, optionally compressed). E.g. I’ve seen packages that had unused .patch files that were not removed after version updates like https://github.com/NixOS/nixpkgs/commit/5bf99d06948c6e6dc7de7dff3d7518d207f442e0.

Also, this is useful for applying patch sets from other derivations, e.g. using fetchurl from Debian (that uses `series` file from quilt, see https://wiki.debian.org/debian/patches and https://wiki.debian.org/UsingQuilt, example: https://salsa.debian.org/debian/bash/-/blob/6a0146056618a32be35ba89e1b092eda2f2fa749/debian/patches/series). This use case currently requires either IFD and custom patch phase.

That is, for simple one-off patches in Nixpkgs, I’d expect maintainers to continue adding individual patch file paths to the `patches` argument, but I think directories make it easier to manage larger sets of patches that are required for some projects and allow using patch management tools.

tie committed:
tests.trivial-builders: test applyPatches

tie committed:
applyPatches: move documentation to the Nixpkgs manual

tie committed:
gnupatch: fix segfault on cleanup

See https://savannah.gnu.org/bugs/?57717

Reproducer:

```console
$ patch -p1 <<'EOF'
diff --git a/file2.txt b/file2.txt
index e69de29..d20e9cd 100644
--- a/file2.txt
+++ b/file2.txt
@@ -1 +1 @@
-old content
+new content
EOF
```

```
(repeated lines omitted)
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.orig : Too many open files
patch: **** Can't create file file2.txt.origSegmentation fault (core dumped)
```

tie commented:
Pushed a commit to fix https://github.com/NixOS/nixpkgs/pull/335579#issuecomment-2295711934.

This PR ready for review. Let me know if I should open separate PRs for 66082f9814e2085d90b0df4b940664f886523d54 and e245b88996d8bbf6939ab37aeb18f675cbc2e91e.

emilazy commented:
Seems like a reasonable change to me, and the Bash code seems high quality as always. I take it the `series` format is compatible with how it is used in Debian, etc.? I’m curious if we have an immediate use case for that, but it seems like sensible functionality and doesn’t add much weight. I’d rather not see a strong proliferation of maintaining forks of packages as patch sets in the Nixpkgs tree, but to the extent that we’re going to, this seems like a clean way of doing it.

I’m wondering if we should require a `series` file when directories are used. I suspect people are not necessarily going to be disciplined about the filename ordering of various patches and it could cause headaches.

philiptaron reviewed (changes requested):
Requesting changes for suspected inode order nondeterminism in patch application.

I _really_ like most of this patch and I'm letting the directory application thing marinate. Appreciate the discussion.

tie commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -1308,38 +1308,92 @@ unpackPhase() {
+            done
+            patchFiles+=( "${lines[@]::n}" )
+        elif [[ -d $p ]]; then
+            # NB nullglob is set in stdenv.
+            patchFiles+=( "$p"/*.{patch,diff}{,.gz,.bz2,.xz,.lzma} )
```
Bash filename expansions are sorted alphabetically.
https://www.gnu.org/software/bash/manual/html_node/Filename-Expansion.html#:~:text=If%20one%20of%20these%20characters%20appears,matching%20the%20pattern

This is locale dependent (not an issue in Nix sandbox) but not affected by inode order.

emilazy replied:
That might be painful for `nix develop`. We should probably set locale variables for that, I guess…

I’d consider this kind of thing another (weak) argument for requiring a `series` file.

philiptaron replied:
Ugh, locales. This pushes me over the line into series-file-required land.

tie replied:
Just double checking that it is indeed locale-dependent.

- https://git.savannah.gnu.org/cgit/bash.git/tree/pathexp.c?id=142bbdd89e4d5bb62aea4469d1d2c24cf470afd8#n426
- https://git.savannah.gnu.org/cgit/bash.git/tree/lib/sh/stringvec.c?id=142bbdd89e4d5bb62aea4469d1d2c24cf470afd8#n170

xaverdh replied:
you could sort it after the fact?

tie commented on pkgs/tools/text/gnupatch/Abort_when_cleaning_up_fails.patch:
```diff
@@ -0,0 +1,51 @@
+From b7b028a77bd855f6f56b17c8837fc1cca77b469d Mon Sep 17 00:00:00 2001
+From: Andreas Gruenbacher <agruen@gnu.org>
+Date: Fri, 28 Jun 2019 00:30:25 +0200
```
Yes, no releases for 7 years, Debian applies this patch as well.

emilazy replied:
I would be in favour of bumping to a Git revision given the number of fixes since then.

I think we should split out the two commits you mentioned into separate PRs so we can discuss things like that.

philiptaron replied:
Yes, absolutely agreed that a git version is the right call vs. vendoring the accumulated patches.

emilazy replied:
https://git.savannah.gnu.org/cgit/patch.git/commit/?id=dce4683cbbe107a95f1f0d45fabc304acfb5d71a looks like the only potentially‐breaking interface change at a glance, but we can probably handle it. We have so many patches for `patch` right now and five of them are CVEs, we should just bump and nag upstream to cut a release.

(But yeah let’s talk about this on another PR!)

emilazy commented:
Another potential reason to require a `series` file is that not every patch will necessarily have a full header, and ones that are imported from upstream will likely have the headers from the upstream commits without a Nixpkgs‐specific rationale. Ensuring we always have a file where we can attach comments to patches would encourage best practices.

tie commented on doc/stdenv/stdenv.chapter.md:
```diff
@@ -1136,6 +1136,29 @@ Example removing all references to the compiler in the output:
+starting with `#`) are ignored. It is an error if a line starts with either `-`
+or `+`.
+
+When processing a directory without a `series` file, only files with standard
+patch extensions `.patch` and `.diff` will be considered for application,
+optionally with compressed file name extension (e.g. `.patch.gz`).
```
See https://github.com/NixOS/nixpkgs/pull/335579#discussion_r1721978177

I agree with https://github.com/NixOS/nixpkgs/pull/335579#issuecomment-2296789754, I’ll update `applyPatches` to return an error if `series` file does not exist to avoid implicit ordering.

philiptaron replied:
Requiring a series file addresses this possibility definitively. I can get on board with that.

emilazy commented:
> The one downside of the series file is that you can delete the entry in the series and orphan a patch file in the repository. Removing that possibility was one reason that this PR exists, as I understand it.

Perhaps it could error out if there are files in the directory not mentioned in the series file. That would be restrictive, and maybe we’d regret it down the line, but it’d be easier to loosen the constraint than to add it later.

emilazy commented:
I’m ambivalent about the source‐of‐truth thing. It doesn’t feel too different to a Nix file importing another. Being able to use other tools to work with series files might be helpful. One thing this change lets us do that might be cool is adding an entire Debian package’s patches directory to `patches`, `series` and all. But on the other hand, who knows if we actually want to encourage that.

philiptaron commented:
OK, I can say ✅ to this with series file required in the directory case and a check that there aren't excess patches hanging about.

Think that's reasonable?

tie commented:
Not sure about the additional check for excess patches. I’m OK with the idea, but not sure what a potential implementation should do. E.g. directory can have `series` file that use patches from subdirectories, https://salsa.debian.org/js-team/nodejs/-/tree/ae4f839ed39c83815d1da835fbc4852e0e6d9971/debian/patches. Performing a recursive check seems to be expensive.

I think it’s fine to omit the check. It’s a lot easier to miss a patch that should be removed when it’s stored alongside a bunch of other _required_ files that are conditionally included from multiple Nix files. Patch set directory, on the other hand, is functionally equivalent to concatenating multiple patches into single file, and ideally managed with some tooling (be that quilt or manual git format-patch exports).

philiptaron commented:
@tie, this PR is sort of rotting on the vine. How'd you like to proceed here? Is there a way to get to unanimity and everyone saying ✅? Should we focus on getting #337981 through first?

emilazy commented:
(I think you mean another PR)

tie commented:
@philiptaron, sorry for the late reply, I’m taking a small break from open source contributions until October.
Yes, I’d like to complete #337961 before proceeding with this PR.

emilazy commented:
Take care; Nixpkgs will be here when you get back, :)
